source/common/network/connection_impl_base.cc

```diff
@@ -50,8 +50,12 @@ void ConnectionImplBase::initializeDelayedCloseTimer() {
 
 void ConnectionImplBase::raiseConnectionEvent(ConnectionEvent event) {
   for (ConnectionCallbacks* callback : callbacks_) {
-    // TODO(mattklein123): If we close while raising a connected event we should not raise further
-    // connected events.
+    // If a previous connected callback closed the connection, don't raise any further connected
+    // events. There was already recursion raising closed events.
+    if (event == ConnectionEvent::Connected && state() != State::Open) {
```

Comment:
cc @danzh2010 I think this will be a semantic merge conflict with your ongoing change FYI.

Reply:
should this change be applied to any non-close event? instead of `event == ConnectionEvent::Connected`, using `event != ConnectionEvent::LocalClosed && event != ConnectionEvent::RemoteClosed`

Reply:
Sure happy to make that change. Either way it's pretty fragile. :(